lib/services/downloads_service.dart

```diff
@@ -0,0 +1,1415 @@
+              }
+            } else {
+              _downloadsLogger.info(
+                  "Recieved status event ${event.status} for finalized download ${listener.name}.  Ignoring.");
+            }
+          } else {
+            _downloadsLogger.severe(
+                "Could not determine item for id ${event.task.taskId}, event:${event.toString()}");
+          }
+        });
+      }
+    });
+
+    // Sometimes we temporarily loose connection while the screen is locked.
+    // Try to restart downloads when the user begins interacting again
+    AppLifecycleListener(onRestart: () {
+      _downloadsLogger
+          .info("App returning from background, restarting downloads.");
+      restartDownloads();
+    }, onHide: () {
+      _showConnectionMessage = false;
+    }, onShow: () {
+      _showConnectionMessage = true;
+    }, onPause: () {
+      _downloadsLogger.info("App is being paused by OS.");
+    });
+
+    FileDownloader().requireWiFi(
+        FinampSettingsHelper.finampSettings.requireWifiForDownloads
+            ? RequireWiFi.forAllTasks
+            : RequireWiFi.forNoTasks);
+
+    bool oldOffline = FinampSettingsHelper.finampSettings.isOffline;
+    bool oldRequireWifi =
+        FinampSettingsHelper.finampSettings.requireWifiForDownloads;
+    FinampSettingsHelper.finampSettingsListener.addListener(() {
+      var newOffline = FinampSettingsHelper.finampSettings.isOffline;
+      var newRequireWifi =
+          FinampSettingsHelper.finampSettings.requireWifiForDownloads;
+      if (oldOffline && !newOffline) {
+        restartDownloads();
+      }
+      if (oldRequireWifi != newRequireWifi) {
+        FileDownloader().requireWiFi(
+            newRequireWifi ? RequireWiFi.forAllTasks : RequireWiFi.forNoTasks);
+      }
+      oldOffline = newOffline;
+      oldRequireWifi = newRequireWifi;
+    });
+  }
+
+  final _downloadsLogger = Logger("downloadsService");
+  final _isar = GetIt.instance<Isar>();
+
+  final _anchor = DownloadStub.fromId(
+      id: "Anchor", type: DownloadItemType.anchor, name: null);
+  late final downloadTaskQueue = IsarTaskQueue(this);
+  late final deleteBuffer = DownloadsDeleteService(this);
+  late final syncBuffer = DownloadsSyncService(this);
+
+  // These track total downloads for the overview on the downloads screen
+  final Map<DownloadItemState, int> downloadStatuses = {};
+  late final Stream<Map<DownloadItemState, int>> downloadStatusesStream;
+  final StreamController<Map<DownloadItemState, int>>
+      _downloadStatusesStreamController = StreamController.broadcast();
+
+  // This triggers refresh of music/artist screens on item deletion
+  late final Stream<void> offlineDeletesStream;
+  final StreamController<void> _offlineDeletesStreamController =
+      StreamController.broadcast();
+
+  // These track node counts for the overview on the downloads screen
+  final Map<String, int> downloadCounts = {"repair": 0};
```

Comment:
No, there isn't.  Repairing is a global state, which is why I just put it in the same map as the song/image count and the syncing nodes count to pass to the downloads_overview.

Reply:
I did notice that while reading the rest of the file, but maybe calling it `repairStep` or something similar is a better idea, since it's not actually counting items/nodes like the other categories? And using a constant instead of making sure the string is the same in all places would be nice!